peglib.h

```diff
@@ -47,6 +47,112 @@ static void* enabler = nullptr; // workaround for Clang version <= 5.0.0
+ *  UTF-8 utilities
+ *---------------------------------------------------------------------------*/
+
+// wchar works differently on linux and windows so everything was done manually
+typedef int codepoint;
+enum class Encoding { ASCII, ANSI, UTF8 }; // ANSI should be called ISO8859 but ANSI is more common and catchy
+
+size_t get_char(const char* s, size_t n, Encoding enc, codepoint& code)
+{
+    if (n < 1) {
+        code = -1;
+        return static_cast<size_t>(-1);
+    }
+
+    const unsigned char* us = reinterpret_cast<const unsigned char*>(s);
+
+    switch(enc) {
+
+    case Encoding::ASCII:
+
+        if (us[0] <= 0x7F) {
+            code = us[0];
+            return 1;
+        }
+        else {
+            code = -1;
+            return static_cast<size_t>(-1);
+        }
+        break;
+
+    case Encoding::ANSI:
+
+        code = us[0];
+        return 1;
+        break;
+
+    case Encoding::UTF8:
+
+        if (us[0] <= 0x7F) { // 0xxx xxxx is ascii
+            code = us[0];
+            return 1;
+        }
+
+        if (us[0] <= 0xBF) { // 10xx xxxx is invalid as first UTF8 byte
+            code = -1;
+            return static_cast<size_t>(-1);
+        }
+
+        if (us[0] <= 0xCF) { // 110x xxxx, 10xx xxxx
+            if (n < 2
+                || us[1] < 0x80 || us[1] > 0xBF)
+            { // second byte missing or invalid
+                code = -1;
+                return static_cast<size_t>(-1);
+            }
+
+            code = (int(us[0] & 0x1F) << 6)
+                + int(us[1] & 0x3F);
+            return 2;
+        }
+
+        if (us[0] <= 0xEF) { // 1110 xxxx, 10xx xxxx, 10xx xxxx
+            if (n < 3
+                || us[1] < 0x80 || us[1] > 0xBF
+                || us[2] < 0x80 || us[2] > 0xBF)
+            { // second or third byte missing or invalid
+                code = -1;
+                return static_cast<size_t>(-1);
+            }
+
+            code = (int(us[0] & 0x0F) << 12)
+                + (int(us[1] & 0x3F) << 6)
+                +  int(us[2] & 0x3F);
+            return 3;
+        }
+
+        if (us[0] <= 0xF7) { // 1111 0xxx, 10xx xxxx, 10xx xxxx, 10xx xxxx
+            if (n < 4
+                || us[1] < 0x80 || us[1] > 0xBF
+                || us[2] < 0x80 || us[2] > 0xBF
+                || us[3] < 0x80 || us[3] > 0xBF)
+            { // second, third or fourth byte missing or invalid
+                code = -1;
+                return static_cast<size_t>(-1);
+            }
+
+            code = (int(us[0] & 0x07) << 18)
+                 + (int(us[1] & 0x3F) << 12)
+                 + (int(us[2] & 0x3F) << 6)
+                 +  int(us[3] & 0x3F);
+            if (code > 0x10FFFF) { // invalid codepoint
+                code = -1;
+                return static_cast<size_t>(-1);
+            }
+            return 4;
+        }
+        break;
+
+    }
+
+    // unknown encoding
+    code = -1;
+    return static_cast<size_t>(-1);
+}
+
```

Comment:
The get_char() function reads one encoded character (which might be several bytes long) from a string, returns the length in bytes and stores the character code in the reference parameter code. Since unicode support works different on windows and linux and all the libraries I found are bloated beyond hope, I have written this myself according to Wikipedia. There is a lot of bit juggling involved and I already fixed many bugs in it. A thorough review and extensive testing are probably necessary.